### PR TITLE
[14.0][ADD] module: product_route_profile

### PR DESCRIPTION
**Pull request by Kev-Roche**
This module replaces the initial concept of route_ids with a new concept of "route profile", coming with a company-specific and priority route profile.

### CODE REVIEW

File: product_route_profile/models/route_profile.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# Copyright 2022 Akretion (https://www.akretion.com).
+# @author Kévin Roche <kevin.roche@akretion.com>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models
+
+
+class RouteProfile(models.Model):
+    _name = "route.profile"
+
```
Comment (CasVissers-360ERP) [inline product_route_profile/models/route_profile.py:11]: Missing description

**Review by CasVissers-360ERP — CHANGES_REQUESTED** [review]
One minor feedback point, we will start our functional tests now. Nice module!

File: product_route_profile/models/route_profile.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    name = fields.Char("Name")
+    company_id = fields.Many2one(
+        comodel_name="res.company",
+        default=lambda self: self.env.company.id,
+        required=False,
+        string="Company",
+    )
+    route_ids = fields.Many2many(
+        "stock.location.route",
+        string="Routes",
```
Comment (florian-dacosta) [inline product_route_profile/models/route_profile.py:21]: I think, for consistency, you should add the domain : `domain=[('product_selectable', '=', True)]` like it is on the original `route_ids` field on template

**Review by CasVissers-360ERP — APPROVED** [review]
Functional review, some code feedback. But module works great. Nice one!

File: product_route_profile/models/route_profile.py (hunk 2 above)
Comment (Kev-Roche) [reply product_route_profile/models/route_profile.py:21]: Good point ! I add it.

File: product_route_profile/models/route_profile.py (hunk 1 above)
Comment (Kev-Roche) [reply product_route_profile/models/route_profile.py:11]: Done, thank you !

File: product_route_profile/models/product_template.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        "route profile, only for this company.",
+    )
+
+    route_ids = fields.Many2many(
+        compute="_compute_route_ids",
+        store=False,
+    )
+
+    @api.depends("route_profile_id", "force_route_profile_id")
+    @api.depends_context("company")
```
Comment (rousseldenis) [inline product_route_profile/models/product_template.py:29]: Where does it depend on this ?

File: product_route_profile/models/product_template.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        "route.profile",
+        string="Priority Route Profile",
+        company_dependent=True,
+        help="If defined, the "
+        "priority route profile will be used and will replace the "
+        "route profile, only for this company.",
+    )
+
+    route_ids = fields.Many2many(
+        compute="_compute_route_ids",
```
Comment (rousseldenis) [inline product_route_profile/models/product_template.py:22]: Maybe a search method should be a great improvement

File: product_route_profile/models/product_template.py (hunk 3 above)
Comment (florian-dacosta) [reply product_route_profile/models/product_template.py:29]: `force_route_profile_id` is company dependent

Comment (rousseldenis) [reply product_route_profile/models/product_template.py:29]: Yes, saw it. But doesn't Odoo ORM aware of that through that field ?

Your new depends IMHO should not add that entry.

**Comment by rousseldenis** [comment]
@Kev-Roche @florian-dacosta IMHO, we should avoid breaking standard behaviour without an option to activate the feature.

At least you should extract this module from tests to do it separately

File: product_route_profile/models/product_template.py (hunk 3 above)
Comment (Kev-Roche) [reply product_route_profile/models/product_template.py:29]: After some checks, it appears to be unnecessary, I remove it.

File: product_route_profile/models/product_template.py (hunk 4 above)
Comment (Kev-Roche) [reply product_route_profile/models/product_template.py:22]: True, I add it to the roadmap

**Comment by Kev-Roche** [comment]
> @Kev-Roche @florian-dacosta IMHO, we should avoid breaking standard behaviour without an option to activate the feature.
> 
> At least you should extract this module from tests to do it separately

I extract the tests from the others.

From my point of view, if we install this module, it is for use it. We prefer to avoid the activation / setting step if you don't mind.

File: product_route_profile/models/product_template.py (hunk 3 above)
Comment (Kev-Roche) [reply product_route_profile/models/product_template.py:29]: After deeper investigation, @api.depends_context("company") is relevant here, as we can see with the fail test (previous commit) and the fix in the last commit.

**Comment by rousseldenis** [comment]
> > @Kev-Roche @florian-dacosta IMHO, we should avoid breaking standard behaviour without an option to activate the feature.
> > At least you should extract this module from tests to do it separately
> 
> I extract the tests from the others.
> 
> From my point of view, if we install this module, it is for use it. We prefer to avoid the activation / setting step if you don't mind.

IMHO, the best is to control features through parameters. Moreover, if it breaks standard behaviour.

Ok, to put in roadmap, but keep it in mind, you'll see it could help you especially if you want to deploy it without activating the feature yet.

File: product_route_profile/models/product_template.py (hunk 3 above)
Comment (rousseldenis) [reply product_route_profile/models/product_template.py:29]: Ok, I think this is a problem in ORM as company_dependant fields that are in depends should imply that behaviour. But ok

**Comment by sebastienbeau** [comment]
@rousseldenis I think adding an option to activate the feature will complexify the code, for something you do not want as real feature (only need it for solving testing compatibility).
So it better/simplier to test the module separately

**Comment by rousseldenis** [comment]
> @rousseldenis I think adding an option to activate the feature will complexify the code, for something you do not want as real feature (only need it for solving testing compatibility). So it better/simplier to test the module separately

In this case, if you don't want an option, shouldn't route_ids field on product have an inverse method in order to set the correct profile (or to create one ?) ?

**Comment by Kev-Roche** [comment]
@rousseldenis I add the search method as previously discussed, you are right, it is very useful. 
About your last comment, we can set and create a route profile (in inventory>settings>Route profiles) and then set the profile in route_profile_id and/or force_route_profile_id in the product template. In that way, route_ids will be equal to the (force)route_profile_id.
We also add a hook to make compatible previous route_ids with the new concept of route profile.

File: product_route_profile/models/product_template.py [hunk 5]
```diff
@@ -32,3 +34,13 @@ def _compute_route_ids(self):
                 rec.route_ids = [(6, 0, self.route_profile_id.route_ids.ids)]
             else:
                 rec.route_ids = False
+
+    def _search_route_ids(self, operator, value):
+        profiles = self.env["route.profile"].search([("route_ids", operator, value)])
```
Comment (rousseldenis) [inline product_route_profile/models/product_template.py]: You can avoid a big search doing this directly in the return:

`
return [
"|",
            ("force_route_profile_id.route_ids", operator, value),
            "&",
            ("force_route_profile_id", "=", False),
            ("route_profile_id.route_ids", operator, value),
]
`

Comment (Kev-Roche) [reply product_route_profile/models/product_template.py]: done, I also squashed the commits.

**Comment by rousseldenis** [comment]
@Kev-Roche Could you fix this ?

**Comment by Kev-Roche** [comment]
@rousseldenis , @sebastienbeau, @CasVissers-360ERP, @florian-dacosta 
I fixed the tests, can we add this module ? 
fixes : 
- add user in addition to administrator access right
- delete first test since it was not really relevelant here and most importantly, the test was inconsistant with some modules (add "default route" on product.template and consequently create a route_profile_id).

If ok, i'll squash the commits.

**Review by CasVissers-360ERP — APPROVED** [review]
LGTM!

**Review by florian-dacosta — APPROVED** [review]
LGTM

**Comment by rousseldenis** [comment]
@Kev-Roche Could you remove travis file as we use Github actions?

**Comment by Kev-Roche** [comment]
> 
@rousseldenis, if I remove the travis file, the tests will not work anymore, are they ?

**Comment by rousseldenis** [comment]
> > 
> 
> @rousseldenis, if I remove the travis file, the tests will not work anymore, are they ?

Of course they do. If you look at checks, both are running :smiley:

**Comment by Kev-Roche** [comment]
> > > 
> > 
> > 
> > @rousseldenis, if I remove the travis file, the tests will not work anymore, are they ?
> 
> Of course they do. If you look at checks, both are running smiley

good point,  I also squash the commits

**Review by rousseldenis — APPROVED** [review]

File: product_route_profile/readme/USAGE.rst [hunk 6]
```diff
@@ -0,0 +1,15 @@
+
+Settings
+=====================
```
Comment (rousseldenis) [inline product_route_profile/readme/USAGE.rst]:
```suggestion
Settings
========
```

Comment (rousseldenis) [reply product_route_profile/readme/USAGE.rst]: @Kev-Roche Could you check titles?

Comment (Kev-Roche) [reply product_route_profile/readme/USAGE.rst]: Done

**Comment by elvise** [comment]
@rousseldenis is it ready for merge ?

**Comment by rousseldenis** [comment]
@Kev-Roche you should run oca-genaddon-readme locally to solve errors.

**Pull request merged**